File: hemlock/hemlock-runtime/docker/agent-identity-kit/node/src/knowledge/semantic.js
import fs from "fs/promises";
import path from "path";

let Pipeline = null;
try {
  const transformers = await import("@xenova/transformers");
  Pipeline = transformers.pipeline;
} catch {
  // Optional dependency — semantic search disabled if not installed
}

export class SemanticSearch {
  constructor(workspace) {
    this.workspace = workspace || process.cwd();
    this.vectorsDir = path.join(this.workspace, "knowledge", "vectors");
    this.embedder = null;
    this.vectors = new Map();
  }

  async init() {
    await fs.mkdir(this.vectorsDir, { recursive: true });
    try {
      const raw = await fs.readFile(path.join(this.vectorsDir, "vectors.json"), "utf-8");
      const data = JSON.parse(raw);
      for (const [id, entry] of Object.entries(data)) {
        this.vectors.set(id, entry);
      }
    } catch {}
  }

  async loadEmbedder() {
    if (this.embedder) return this.embedder;
    if (!Pipeline) {
      throw new Error("Semantic search requires @xenova/transformers. Install with: npm install @xenova/transformers");
    }
    this.embedder = await Pipeline("feature-extraction", "Xenova/all-MiniLM-L6-v2");
    return this.embedder;
  }

  async embed(text) {
    const pipe = await this.loadEmbedder();
    const output = await pipe(text, { pooling: "mean", normalize: true });
    return Array.from(output.data);
  }

  cosineSimilarity(a, b) {
    let dot = 0, normA = 0, normB = 0;
    for (let i = 0; i < a.length; i++) {
      dot += a[i] * b[i];
      normA += a[i] * a[i];
      normB += b[i] * b[i];
    }
    return dot / (Math.sqrt(normA) * Math.sqrt(normB));
  }

  async indexDocument(docId, text, metadata = {}) {
    const chunks = [];
    const chunkSize = 3500;
    const overlap = 350;

    if (text.length <= chunkSize) {
      chunks.push({ id: docId, text, metadata });
    } else {
      let start = 0;
      let i = 0;
      while (start < text.length) {
        const end = Math.min(start + chunkSize, text.length);
        chunks.push({
          id: `${docId}:chunk-${i}`,
          text: text.slice(start, end),
          metadata: { ...metadata, parent: docId, chunkIndex: i },
        });
        start = end - overlap;
        i++;
        if (start + overlap >= text.length) break;
      }
    }

    for (const chunk of chunks) {
      const embedding = await this.embed(chunk.text);
      this.vectors.set(chunk.id, {
        embedding,
        text: chunk.text.slice(0, 200),
        metadata: chunk.metadata,
      });
    }

    return chunks.length;
  }

  async save() {
    const data = {};
    for (const [id, entry] of this.vectors) {
      data[id] = entry;
    }
    await fs.writeFile(
      path.join(this.vectorsDir, "vectors.json"),
      JSON.stringify(data)
    );
  }

  async search(query, limit = 10) {
    const queryEmbedding = await this.embed(query);
    const results = [];

    for (const [id, entry] of this.vectors) {
      const score = this.cosineSimilarity(queryEmbedding, entry.embedding);
      results.push({
        id,
        score,
        text: entry.text,
        metadata: entry.metadata,
      });
    }

    results.sort((a, b) => b.score - a.score);
    return results.slice(0, limit);
  }

  async hybridSearch(query, keywordResults, limit = 10) {
    const semanticResults = await this.search(query, limit * 2);

    const rrf = new Map();
    const k = 60;

    for (let i = 0; i < keywordResults.length; i++) {
      const id = keywordResults[i].docId || keywordResults[i].id;
      const score = 1 / (k + i + 1);
      rrf.set(id, (rrf.get(id) || 0) + score);
    }

    for (let i = 0; i < semanticResults.length; i++) {
      const id = semanticResults[i].metadata?.parent || semanticResults[i].id;
      const score = 1 / (k + i + 1);
      rrf.set(id, (rrf.get(id) || 0) + score);
    }

    const sorted = [...rrf.entries()].sort((a, b) => b[1] - a[1]);
    return sorted.slice(0, limit).map(([id, score]) => ({
      id,
      rrfScore: score,
      text: this.vectors.get(id)?.text || "",
      metadata: this.vectors.get(id)?.metadata || {},
    }));
  }

  status() {
    return {
      vectors: this.vectors.size,
      semanticAvailable: !!Pipeline,
      vectorsDir: this.vectorsDir,
    };
  }
}

File: hemlock/hemlock-runtime/docker/agent-identity-kit/python/agent_identity_kit/semantic.py
"""
Semantic search — vector embeddings over indexed chunks, with hybrid (semantic +
keyword) ranking via Reciprocal Rank Fusion.

Mirrors the Node `src/knowledge/semantic.js`. Fully optional: the embedding model
(`@xenova/transformers` in Node, `transformers`/sentence-transformers in Python) is
loaded lazily and only when actually needed, so this component is bypassed entirely
if you don't use it.
"""

from __future__ import annotations
import json
import math
from pathlib import Path
from datetime import datetime, timezone


class SemanticSearch:
    def __init__(self, workspace: str):
        self.workspace = str(workspace)
        self.vectors_dir = Path(workspace) / "knowledge" / "vectors"
        self.embedder = None
        self.vectors = {}

    def init(self):
        self.vectors_dir.mkdir(parents=True, exist_ok=True)
        try:
            raw = (self.vectors_dir / "vectors.json").read_text(encoding="utf-8")
            self.vectors = json.loads(raw)
        except (FileNotFoundError, json.JSONDecodeError):
            self.vectors = {}

    async def load_embedder(self):
        if self.embedder:
            return self.embedder
        # Lazy, optional dependency.
        from sentence_transformers import SentenceTransformer
        self.embedder = SentenceTransformer("all-MiniLM-L6-v2")
        return self.embedder

    def _embed_sync(self, text):
        """Synchronous embedding (sentence-transformers is sync)."""
        import asyncio
        try:
            loop = asyncio.get_event_loop()
        except RuntimeError:
            loop = asyncio.new_event_loop()
        return loop.run_until_complete(self.embed(text))

    async def embed(self, text):
        model = await self.load_embedder()
        vec = model.encode(text, normalize_embeddings=True)
        return list(vec)

    @staticmethod
    def cosine_similarity(a, b):
        dot = sum(x * y for x, y in zip(a, b))
        na = math.sqrt(sum(x * x for x in a))
        nb = math.sqrt(sum(x * x for x in b))
        if na == 0 or nb == 0:
            return 0.0
        return dot / (na * nb)

    def index_document(self, doc_id, text, metadata=None):
        metadata = metadata or {}
        chunk_size, overlap = 3500, 350
        chunks = []
        if len(text) <= chunk_size:
            chunks.append({"id": doc_id, "text": text, "metadata": metadata})
        else:
            start = 0
            i = 0
            while start < len(text):
                end = min(start + chunk_size, len(text))
                chunks.append({"id": f"{doc_id}:chunk-{i}", "text": text[start:end],
                               "metadata": {**metadata, "parent": doc_id, "chunkIndex": i}})
                start = end - overlap
                i += 1
                if start + overlap >= len(text):
                    break
        for ch in chunks:
            self.vectors[ch["id"]] = {
                "embedding": self._embed_sync(ch["text"]),
                "text": ch["text"][:200],
                "metadata": ch["metadata"],
            }
        return len(chunks)

    def save(self):
        (self.vectors_dir / "vectors.json").write_text(json.dumps(self.vectors))

    def search(self, query, limit=10):
        q = self._embed_sync(query)
        results = []
        for cid, entry in self.vectors.items():
            results.append({"id": cid, "score": self.cosine_similarity(q, entry["embedding"]),
                            "text": entry["text"], "metadata": entry["metadata"]})
        results.sort(key=lambda r: r["score"], reverse=True)
        return results[:limit]

    def hybrid_search(self, query, keyword_results, limit=10):
        semantic = self.search(query, limit * 2)
        rrf = {}
        k = 60
        for i, kr in enumerate(keyword_results):
            rid = kr.get("docId") or kr.get("id")
            rrf[rid] = rrf.get(rid, 0) + 1 / (k + i + 1)
        for i, sr in enumerate(semantic):
            rid = sr["metadata"].get("parent") or sr["id"]
            rrf[rid] = rrf.get(rid, 0) + 1 / (k + i + 1)
        sorted_ids = sorted(rrf.items(), key=lambda x: x[1], reverse=True)[:limit]
        out = []
        for rid, score in sorted_ids:
            entry = self.vectors.get(rid, {})
            out.append({"id": rid, "rrfScore": score, "text": entry.get("text", ""),
                        "metadata": entry.get("metadata", {})})
        return out

    def status(self):
        try:
            from sentence_transformers import SentenceTransformer  # noqa
            available = True
        except ImportError:
            available = False
        return {"vectors": len(self.vectors), "semanticAvailable": available,
                "vectorsDir": str(self.vectors_dir)}
